source/common/network/connection_impl.cc

```diff
@@ -785,9 +785,14 @@ void ConnectionImpl::dumpState(std::ostream& os, int indent_level) const {
 ServerConnectionImpl::ServerConnectionImpl(Event::Dispatcher& dispatcher,
                                            ConnectionSocketPtr&& socket,
                                            TransportSocketPtr&& transport_socket,
-                                           StreamInfo::StreamInfo& stream_info, bool connected)
+                                           StreamInfo::StreamInfo& stream_info, bool connected,
+                                           Buffer::Instance* inspect_buffer)
     : ConnectionImpl(dispatcher, std::move(socket), std::move(transport_socket), stream_info,
-                     connected) {}
+                     connected) {
+  if (inspect_buffer) {
+    getReadBuffer().buffer.move(*inspect_buffer);
```

Comment:
The initial data needs to be put where the transport socket reads from. 

Example: if this connection used the TLS transport socket, this code would put ciphertext (TLS messages) into the read buffer for Network Filters, not plaintext.

Reply:
I thought the plaintext will go through the TLS transport socket again(maybe in the next onRead() cb is triggered), anyway, it is my guess, I will dig into it.